File: src/inJava/inheritance/WithComposition2.java
package inJava.inheritance;

import static java.lang.System.out;

// Using composition and interfaces to recoup both
// code reuse *and* polymorphism.
class WithComposition2 {
    static void compositionProgram(){
        Pet pet = new Pet("any",0);
        Dog dog = new Dog("Lassie",4);
        Cat cat = new Cat("Misse",5,false);
        // Still can't do this:
        // pet = dog;
        // But we can now do this:
        CanSpeak aSpeakingPet = dog;

        // Use delegated methods
        out.println(dog.say());
        out.println(aSpeakingPet.say());
        out.println(cat.say());
    }

    static interface CanSpeak {
        String say();
    }

    static interface IsPet {
        String getName();

        int getAge();
    }


    static class Pet implements IsPet {
        private final String name;
        private final int age;

        protected Pet(String name, int age) {
            // ... and initialize their values inside constructor.
            this.name = name;
            this.age = age;
        }

        public String getName(){
            return this.name;
        }

        public int getAge() {
            return this.age;
        }
    }

    static class Dog implements CanSpeak, IsPet {
        // Dog is still an extension of Pet - not a subclass, we extend it manually
        // by *composing* it with an object of Pet, the methods of which we
        // can then use.
        private final Pet superPet;
        public Dog(String name, int age){
            this.superPet = new Pet(name, age);
        }

        // get_name() and get_age() implemented through *delegation*,
        // i.e. passing the call on to the helper class.
        public String getName() {
            return superPet.getName();
        }

        public int getAge() {
            return superPet.getAge();
        }

        @Override
        public String say() {
            return getName() + " says Arf";
        }
    }

    static class Cat implements CanSpeak, IsPet {
        private final WithComposition1.Pet superPet;
        boolean isEvil;

        public Cat(String name, int age, boolean isEvil){
            this.superPet = new WithComposition1.Pet(name, age);
        }

        // get_name() and get_age() implemented through *delegation*,
        // i.e. passing the call on to the helper class.
        public String getName() {
            return superPet.getName();
        }

        public int getAge() {
            return superPet.getAge();
        }

        @Override
        public String say() {
            return getName() + " says Meow";
        }

        boolean isEvil(){ return this.isEvil; }

    }

    // The Java "invocation" to make a class executable
    public static void main(String[] args) {
        compositionProgram();
    }
}

File: src/inPython/inheritance/WithComposition2.py
# package python.inheritance

# Using composition and pure interfaces to recoup both
# code reuse and polymorphism.
from abc import ABC, abstractmethod


def inheritance_program():
    # pet = Pet("any", 0)     # Can't do this any more, Pet is abstract

    dog = Dog("Lassie", 4)
    cat = Cat("Misse", 5, False)
    # Still can't do this without a warning:
    pet : Pet = dog
    # But we can now do this:
    a_speaking_pet: CanSpeak = dog

    # Use overridden methods
    print(dog.say())
    print(a_speaking_pet.say())
    print(cat.say())


class CanSpeak(ABC):
    @abstractmethod
    def say(self):
        raise NotImplementedError


class IsPet(ABC):
    @abstractmethod
    def get_name(self):
        raise NotImplementedError

    def get_age(self):
        raise NotImplementedError


class Pet(IsPet, ABC):
    def __init__(self, name: str, age: int):
        self.name = name  # Everything common for Cat and Dog here
        self.age = age

    def get_name(self):
        return self.name

    def get_age(self):
        return self.age


class Dog(CanSpeak, IsPet):
    def __init__(self, name: str, age: int):
        self.__pet = Pet(name, age)

    # Must implement say() since Dog is a non-abstract class
    def say(self):
        return f"{self.get_name()} says Arf"

    def get_name(self):
        return self.__pet.get_name()

    def get_age(self):
        return self.__pet.get_age()


class Cat(CanSpeak, IsPet):
    def __init__(self, name: str, age: int, is_evil: bool):
        self.__pet = Pet(name, age)
        self.is_evil = is_evil

    # Must implement say() since Cat is a non-abstract class
    def say(self):
        return f"{self.get_name()} says Meow"

    def get_name(self):
        return self.__pet.get_name()

    def get_age(self):
        return self.__pet.get_age()

    # get_name() and get_age() inherited but invisible

    def is_evil(self):
        return self.is_evil


if __name__ == "__main__":
    inheritance_program()
